### PR TITLE
Python 3 compatibility via compat.py

### PR DESCRIPTION
**Pull request by cclauss**
Fixes for 30 flake8 errors on Python 3...

flake8 testing of https://github.com/kensho-technologies/graphql-compiler on Python 3.6.2

$ __flake8 . --count --select=E901,E999,F821,F822,F823 --show-source --statistics__

```
./graphql_compiler/debugging_utils.py:19:18: F821 undefined name 'xrange'
        for i in xrange(0, len(too_many_parts) - 1, 2)
                 ^

./graphql_compiler/debugging_utils.py:24:14: F821 undefined name 'xrange'
    for i in xrange(1, len(parts)):
             ^

./graphql_compiler/debugging_utils.py:87:26: F821 undefined name 'xrange'
                for i in xrange(1, len(separate_keywords) - 1, 2):
                         ^

./graphql_compiler/compiler/blocks.py:55:35: F821 undefined name 'basestring'
                all(isinstance(x, basestring) for x in self.start_class)):
                                  ^

./graphql_compiler/compiler/blocks.py:98:35: F821 undefined name 'basestring'
                all(isinstance(x, basestring) for x in self.target_class)):
                                  ^

./graphql_compiler/compiler/blocks.py:256:43: F821 undefined name 'basestring'
        if not isinstance(self.direction, basestring):
                                          ^

./graphql_compiler/compiler/blocks.py:320:43: F821 undefined name 'basestring'
        if not isinstance(self.direction, basestring):
                                          ^

./graphql_compiler/compiler/blocks.py:347:22: F821 undefined name 'xrange'
            for i in xrange(self.depth + 1)
                     ^

./graphql_compiler/compiler/expressions.py:78:35: F821 undefined name 'basestring'
        if isinstance(self.value, basestring):
                                  ^

./graphql_compiler/compiler/expressions.py:102:37: F821 undefined name 'basestring'
        elif isinstance(self.value, basestring):
                                    ^

./graphql_compiler/compiler/expressions.py:107:36: F821 undefined name 'basestring'
            elif all(isinstance(x, basestring) for x in self.value):
                                   ^

./graphql_compiler/compiler/expressions.py:182:41: F821 undefined name 'unicode'
        match_variable_name = '{%s}' % (unicode(variable_with_no_dollar_sign),)
                                        ^

./graphql_compiler/compiler/expressions.py:206:20: F821 undefined name 'unicode'
            return unicode(self.variable_name)
                   ^

./graphql_compiler/compiler/expressions.py:237:16: F821 undefined name 'unicode'
        return unicode(self.field_name)
               ^

./graphql_compiler/compiler/expressions.py:619:42: F821 undefined name 'unicode'
        if not isinstance(self.operator, unicode):
                                         ^

./graphql_compiler/compiler/helpers.py:10:36: F821 undefined name 'unicode'
VARIABLE_ALLOWED_CHARS = frozenset(unicode(string.ascii_letters + string.digits + '_'))
                                   ^

./graphql_compiler/compiler/helpers.py:60:30: F821 undefined name 'basestring'
    if not isinstance(value, basestring):
                             ^

./graphql_compiler/compiler/helpers.py:62:12: F821 undefined name 'unicode'
    return unicode(value)
           ^

./graphql_compiler/compiler/helpers.py:100:30: F821 undefined name 'basestring'
    if not isinstance(value, basestring):
                             ^

./graphql_compiler/compiler/helpers.py:158:44: F821 undefined name 'basestring'
        if field and not isinstance(field, basestring):
                                           ^

./graphql_compiler/compiler/helpers.py:185:34: F821 undefined name 'basestring'
        if not isinstance(child, basestring):
                                 ^

./graphql_compiler/compiler/helpers.py:199:60: F821 undefined name 'unicode'
        mark_name = u'__'.join(self.query_path) + u'___' + unicode(self.visit_counter)
                                                           ^

./graphql_compiler/query_formatting/gremlin_formatting.py:19:30: F821 undefined name 'basestring'
    if not isinstance(value, basestring):
                             ^

./graphql_compiler/query_formatting/gremlin_formatting.py:89:43: F821 undefined name 'basestring'
        if not isinstance(argument_value, basestring):
                                          ^

./graphql_compiler/query_formatting/gremlin_formatting.py:90:30: F821 undefined name 'unicode'
            argument_value = unicode(argument_value)
                             ^

./graphql_compiler/query_formatting/match_formatting.py:18:30: F821 undefined name 'basestring'
    if not isinstance(value, basestring):
                             ^

./graphql_compiler/query_formatting/match_formatting.py:75:43: F821 undefined name 'basestring'
        if not isinstance(argument_value, basestring):
                                          ^

./graphql_compiler/query_formatting/match_formatting.py:76:30: F821 undefined name 'unicode'
            argument_value = unicode(argument_value)
                             ^

./graphql_compiler/tests/test_helpers.py:37:14: F821 undefined name 'xrange'
    for i in xrange(len(expected_blocks)):
             ^

./graphql_compiler/tests/test_ir_generation_errors.py:441:48: F821 undefined name 'xrange'
            args = (variable_names[i] for i in xrange(num_args))
                                               ^
```

### CODE REVIEW

File: setup.py [hunk 1]
```diff
@@ -27,5 +27,5 @@
           'Programming Language :: Python :: 2.7',
       ],
       keywords='graphql database compiler orientdb',
-      python_requires='>=2.7, <3',
+      python_requires='>=2.7, <3.7',
```
Comment (obi1kenobi) [inline setup.py]: I'd make this `<4` since I'd expect `3.7` won't break the few things this library depends on.

File: graphql_compiler/compat.py [hunk 2]
```diff
@@ -0,0 +1,17 @@
+#!/usr/bin/env/python
+
+"""
+    A minimal compatibility shim for Python 2 and Python 3.
```
Comment (obi1kenobi) [inline graphql_compiler/compat.py]: I'd make this a one-line docstring on line 2: 
```"""A minimal compatibility shim for Python 2 and Python 3."""```

File: .travis.yml [hunk 3]
```diff
@@ -2,12 +2,18 @@ sudo: false
...
 install:
   - pip install -r dev-requirements.txt
   - pip install -r requirements.txt
   - pip install -e .
+before_script:
+  - pip install flake8
+  # stop the build if there are Python syntax errors or undefined names
+  - time flake8 . --count --select=E901,E999,F821,F822,F823 --show-source --statistics
+  # exit-zero treats all errors as warnings.  The GitHub editor is 127 chars wide
+  - time flake8 . --count --exit-zero --max-complexity=10 --max-line-length=127 --statistics
```
Comment (obi1kenobi) [inline .travis.yml]: I assume these are temporary, just while you're getting the PR ready, right?

**Comment by cclauss** [comment]
@obi1kenobi I have done what I could...  Someone who knows the code is going to need to pick up the thread from [here](https://travis-ci.org/kensho-technologies/graphql-compiler/builds/255863885).  Thinks seem fine again on Python 2.7 and almost all the errors in Python 3 seem to about x.\_\_str\_\_() or y.\_\_repo\_\_() returning bytes, not str.  Perhaps str(return_value) will work.

I hope you can squash commits!!!

**Comment by obi1kenobi** [comment]
Hi @cclauss, thank you for the hard work getting the code into good shape. I'd be happy to take over, but before I'm able to do that, would you mind signing our contributor licensing agreement? It's just a standard form that many open-source projects use -- you can find the link in the "checks" section, just click on `Details`.

Thanks again for the help and for the great work!

**Comment by obi1kenobi** [comment]
Hi @cclauss, just a reminder to please sign our contributor licensing agreement (CLA) so I can merge your hard work into the project and get Python 3 support out the door!

Thanks!

**Comment by obi1kenobi** [comment]
Superseded by #31.

**Pull request closed without merging**